### PR TITLE
Logic tests again

### PR DESCRIPTION
**Pull request by mariusandra**
## Changes

Edit: read the last comment. Description is out of date.

TL;DR: Logic tests are looking more and more like a reality. Still a bit more work to do though.

- Based on previous work in https://github.com/PostHog/posthog/pull/4689
- This adds two tests to `infiniteListLogic`, both doing the same, but one using `given` and one in a more verbose style
- Given loses all types, the more verbose version keeps them. 
- Not making conclusions yet as for which is the best approach (or neither), though I still feel more comfortable when the types are there.
- I also added a two test helpers: `mockAPIGet` and `initKeaTestLogic`
- Still WIP!

Ideas to try out:

```ts
// 1. test that specific actions fired in order
expectToHaveRanSinceLastTime([
  logic.actionTypes.loadResults,
  logic.actionTypes.loadResultsSuccess,
  router.actionTypes.push,
])
expectNoActionToHaveBeenDispatchedSinceTheLastTimeYouCalledThisFunction()

// 2. save the log of actions:
expect(getActionsSinceLasRun()).toMatchSnapshot()

// 3. override/mock listeners
logic.listeners.loadResults = () => { ... } // to test
```

The plan is to fully type the taxonomic filter component (not the taxonomic *property* filter) and see from there. So one more logic file. Plus the missing bits still from here.

## Checklist

- [ ] All querysets/queries filter by Organization, by Team, and by User
- [ ] Django backend tests
- [ ] Jest frontend tests
- [ ] Cypress end-to-end tests
- [ ] Migrations are safe to run at scale (e.g. PostHog Cloud) – present proof if not obvious
- [ ] New/changed UI is decent on smartphones (viewport width around 360px)

### CODE REVIEW

**Comment by mariusandra** [comment]
All right, time to bring this in. @macobo you probably won't agree with my conclusion though 😅 

After playing with two approaches, one with `given`, and one without, I'm now convinced that I want to remove given and keep the traditional test. 

My arguments:
- Given loses all types. I'm still 100% *for* writing tests in TypeScript. This way we get a) autocomplete when writing the tests, b) type level check that we're calling the right actions with the right values, c) early warnings if we refactor code and some tests start failing because the type of the third argument to an action has changed. In short: TS has made writing JS so much nicer and I'm not going to abandon that for tests. It has been very straightforward for me to write tests with TS as well, except, obviously, when using `given`.
- Given + mocking makes tests unreadable for me. I'm sorry, but I've tried many times and whenever I see code [like this](https://github.com/PostHog/posthog-js/blob/master/src/__tests__/posthog-core.js#L332-L346) (random example), I can't understand what's happening by looking at the code. I have to trace everything from the first line of the tests to know what values are where and what's happening. This is highly subjective, though I have not been able to shake this feeling for a while now. I want to look at a test and know what it's testing.
- Using the system of `given`s and `mock`s, it feels like there's no way around re-implementing virtually every tiny technical detail from every logic in a test, as it feels everything needs to be mocked, no matter how internal to a class. I fear the only thing this will accomplish is make people think "ugh" when they need to write tests, effectively limiting their effectiveness. A `logic` should be a black box with actions in and values out. We can mock API calls, but IMO we should not need to mock anything else! 
- The actions/values in logic are always "two dots" away, so I also don't understand the write-time convenience of using `given.subject()` instead of `logic.values.counter`. This is compact enough that I feel comfortable just writing `logic.actions.setSearchQuery('event')` instead of `given.logic.actions.setSearchQuery('event')` and `logic.values.searchQuery` instead of `given.subject()` or `given.searchQuery()`. 

This all sounds like a rant, and I'm sure there are various legit reasons to use `given`, and to not use types, etc. Perhaps it only "clicks" if everything is written in this way, and not before... sort of how kea shines if everything is in kea and nothing in `use*` calls. However, we need to make progress here and at the end of the day, I don't see how using a lot of givens and mocks makes it easier for us as a company test more code.

In short: `true` to using TypeScript in tests and `false` to using given.

**Comment by mariusandra** [comment]
I have now implemented some basic tests for `infiniteListLogic`. Some more tests need to be written to consider the taxonomic filter "done", yet we can expand after. I'd like to get these utility methods in though so we could also start writing other logic tests.

---

Things to try some day or when the need arises when writing the next things:

```ts
// 1. test that specific actions fired in order (make a custom redux middleware and add that to `resetContext` that captures the actions)
expectToHaveRanSinceLastTime([
  logic.actionTypes.loadResults,
  logic.actionTypes.loadResultsSuccess,
  router.actionTypes.push,
])
expectNoActionToHaveBeenDispatchedSinceTheLastTimeYouCalledThisFunction()

// 2. save the log of actions:
expect(getActionsSinceLasRun()).toMatchSnapshot()

// 3. override/mock listeners
logic.listeners.loadResults = () => { ... } // to test
```

**Review by paolodamico — APPROVED** [review]
Great stuff! Useful that we're starting to add logic tests. I'm also a fan of having them on TS to spot bugs early. FYI simplified the mock data as I couldn't find a reason to generate so many test elements.
